### DISCUSSION TITLE
Support for Pico W?

### DISCUSSION THREAD

**Discussion by sgbaird** [MrYsLab/telemetrix-rpi-pico-w#3] [General]

**Comment by MrYsLab** [comment]
Thanks for the question. I apologize in advance since this will be a rather long answer to a simple question.

There are two main challenges in providing support for Pico W. 

The first challenge is that I would need to convert the code from using the RPi Pico C SDK to using [Earle F. Philhower's Arduino library](https://github.com/earlephilhower/arduino-pico). 

Adding WiFi support to the current Pico SDK version is not an issue. However, the user will have to edit the Pico library code to provide their network's SSID and Password and recompile it before uploading it to the Pico. The most straightforward development environment to do this would be the Arduino IDE environment. To continue using the RPi Pico C SDK would require the user to set up a rather complex VS code environment, for which I am not willing to provide support.

The second challenge is that I have not been able to find anywhere I can purchase the Pico W locally. If you are willing to buy two units and solder headers and provide them to me, I would be willing to do the port in exchange.

If this sounds acceptable, I propose I try doing the port to a non-wifi version board. If that is successful, I will let you know, and you can then provide the Pico W boards.

Doing the port should not be a considerable challenge (famous last words) since I already have an Arduino Core version for the [Arduino Nano 2040 Connect](https://github.com/MrYsLab/telemetrix-nano-2040-wifi) board. It uses a different board type, but the code should be very similar. With any luck, the Python side should not have to change for the threaded or asyncio versions of the code.

Please let me know your thoughts.

**Comment by sgbaird** [comment]
Hi @MrYsLab thanks for the detailed response!

> The first challenge is that I would need to convert the code from using the RPi Pico C SDK to using [Earle F. Philhower's Arduino library](https://github.com/earlephilhower/arduino-pico).
> 
> Adding WiFi support to the current Pico SDK version is not an issue. However, the user will have to edit the Pico library code to provide their network's SSID and Password and recompile it before uploading it to the Pico. The most straightforward development environment to do this would be the Arduino IDE environment. To continue using the RPi Pico C SDK would require the user to set up a rather complex VS code environment, for which I am not willing to provide support.

Do you mean setting up VS Code is a lot more complex than Arduino IDE (which is almost certainly true), or is setting up the Pico SDK in VS Code pretty complex?

I'm guessing something like Thonny + MicroPython [per this document](https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf) might not work well for some reason?

> The second challenge is that I have not been able to find anywhere I can purchase the Pico W locally. If you are willing to buy two units and solder headers and provide them to me, I would be willing to do the port in exchange.

I recently bought some Pico W's with separate headers [from PiShop](https://www.pishop.us/product/raspberry-pi-pico-w/?src=raspberrypi) that I plan to solder and connect to [Maker Pi Pico Bases](https://www.adafruit.com/product/5160) for an [autonomous research lab demo](https://hackaday.io/project/186289-autonomous-research-laboratories) [[GitHub](https://github.com/sparks-baird/self-driving-lab-demo)]. Out of curiosity, why two units?

> If this sounds acceptable, I propose I try doing the port to a non-wifi version board. If that is successful, I will let you know, and you can then provide the Pico W boards.

Sounds interesting. I'll give this some more thought! I have a conference related to the demo I mentioned at the end of the month, so I'll probably need to look for another solution in the meantime, but I would love to come back to this after that. Do you have any suggestions in the short term for getting some kind of server set up with the RPi Pico W so that I can control the brightness and RGB values of the Maker Pi Pico Base NeoPixel and read sensor data from the AS7341?

**Comment by MrYsLab** [comment]
_Do you mean setting up VS Code is a lot more complex than Arduino IDE (which is almost certainly true), or is setting up the Pico SDK in VS Code pretty complex?_

Setting up VS Code is simple, but getting setup with the SDK, cmake and my code is painful if not impossible. I don't use VS Code, but JetBrains CLion. This is a commercial product, so I would not ask users to purchase it. I have never been able to get VS Code to build my JetBrains code. There is something different in how the two tools handle cmake.

_I'm guessing something like Thonny + MicroPython [per this document](https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf) might not work well for some reason?_

MircoPython is very slow. It also has a problem in that there is a conflict with its REPL and the read function. I documented the conflict (sort of) in [this article](https://mryslab.github.io/bots-in-pieces/raspberry/pi/pico/2021/02/15/pico-adventures-part2.html).

_why two units?_

In case I blow one up.

Let me know if you want me to go ahead.


Thanks,

Alan

**Comment by sgbaird** [comment]
Thanks for the clarifications! Yeah, let's do it!

**Comment by MrYsLab** [comment]
Ok. I will start by porting using my non-wifi pico. If that is successful, I will let you know, and we can go from there. Please don't buy anything yet, in case I am not successful.

**Comment by sgbaird** [comment]
@MrYsLab sounds good!

**Comment by MrYsLab** [comment]
I have [ported the code](https://github.com/MrYsLab/Telemetrix4RpiPicoW) using [Earle F. Philhower's Arduino library](https://github.com/earlephilhower/arduino-pico). I have a lot of testing to do but was able to blink the onboard LED, so I know I have serial communication functioning. I am running the non-wifi version of the board. I want to test all functionality using this library before asking you to purchase the pico W.

There is neopixel support, as well as AccelStepper stepper motor support. The grove ports should be supported since they are attached to standard GPIO pins.

I will post again when testing is complete.

**Comment by MrYsLab** [comment]
Yet another quick update. Going through Philhower's library in some detail, it supports Pico features that the Arduino Nano RP2040 connect does not. Features such as supporting both i2c and spi ports and not just one. There may be others I have not uncovered as of yet.

What does this all boil down to? I already support these features for the non-wifi pico and want to support them for the pico W. It will take me a little time to incorporate Philhower's implementation and test it. Stay tuned.

**Comment by MrYsLab** [comment]
@sgbaird I moved your original issue to this discussion, and I am keeping track of my progress [here](https://github.com/MrYsLab/telemetrix-rpi-pico-w/discussions/1).

**Comment by MrYsLab** [comment]
@sgbaird Question: Does your application require SPI support? I am thinking of not including it for the Pico W. I am not sure if my mpu9250 has gone bad or not, but I can't get it to work on any hardware platforms.

**Reply by sgbaird** [reply]
I've only used I2C so far for the project.

**Comment by MrYsLab** [comment]
Thanks. I only have one interface left (besides spi): stepper motor support. I hope to test that this weekend.

**Reply by MrYsLab** [reply]
@sgbaird I've been under the weather (non-covid) for the last couple of days. I forgot that I did not implement stepper motor support for the Pico. I think it should work "out of the box for the Pico W," but if not, I can leave it out. I received a BME280 sensor for SPI. Hopefully, I should be able to get that going with a little effort. 

That being said, I believe I am ready to start implementing the WIFI interface, but I will need a Pico W to do so. If you are still interested in providing the devices, please email to the address below my avatar on [this page](https://github.com/MrYsLab). I will reply with my physical mailing address. I have tried to purchase the Pico W, but it is still unavailable from the usual suppliers I use.

**Reply by sgbaird** [reply]
Sounds like you've made some excellent progress! Sorry to hear that you've been sick. I just sent you an email; I'm down to just one Pico W, so I'll probably need to order some more. I originally sourced them from PiShop with reasonable prices and shipping fees, but they're out of stock now. They're in stock at CanaKit, but it looks pricey between the header pins' cost and the shipping fees. I was planning to buy more anyway, so I'll keep an eye out.
